Ipa interface overview (#1)
KevinUTAT/oneAPI-samples_wenkaixu_fork: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/IPAuthoring_interfaces_overview/host-pipes/Readme.md

# Host Pipe Streaming

> This example demonstrate a vector addition design using streaming date interface.
> 
> This design will show how to use host pipe to stream data in and out of kernel and how to customize their properties.

Reviewer: ```suggestion
This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to configure a streaming data interface.
```

## Invocation Interface

> By default, an un-decorated oneAPI kernel will have all its control signals and arguments passed through an MM-Agent interface:

Reviewer: ```suggestion
By default, an un-decorated oneAPI kernel will have all its control signals and arguments mapped into the IP component's control/status register (CSR).
```

Also, please include a figure from here (you may need to customize... let me know if you can't find one in here
https://intel-my.sharepoint.com/personal/paul_white_intel_com/Documents/Documentation/HLS%20Docs/kernel-top-levels.vsdx?web=1

> ## Data Interface - Host Pipe

Reviewer: ```suggestion
## Data Interface - Streaming (Host Pipes)
```

> Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design. Pipes that connect a host and a device are referred to as host pipes. In this design, interfaces `A_in`, `B_in`, `C_in` are implemented as Host Pipes.

Reviewer: ```suggestion
Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design. Pipes that connect a host and a device are referred to as host pipes. In this design, interfaces `A_in`, `B_in`, `C_out` are implemented as host pipes.
```

> Host pipe support is enabled by including the following include statement in your design:
> ```cpp
> #include <sycl/ext/intel/experimental/pipes.hpp>
> ```

Reviewer: ```suggestion
```
nope, now it just comes from `fpga_extensions.hpp`

> ### Declear Host Pipe

Reviewer: ```suggestion
### Declare Host Pipe
```

Each individual host pipe is a program scope class declaration of the templated pipe class:
```cpp
sycl::ext::intel::experimental::pipe<id, type, min_capacity, properties>;
```

> The table shows how to configure each templated parameter:

Reviewer: these details should be covered in a specific tutorial or documentation. you don't need all this here.

> - For non-POSIX shells, like csh, use the following command: `bash -c 'source <install-dir>/setvars.sh ; exec csh'`
>
> Windows*:
> - `C:\Program Files(x86)\Intel\oneAPI\setvars.bat`
> - Windows PowerShell*, use the following command: `cmd.exe "/K" '"C:\Program Files (x86)\Intel\oneAPI\setvars.bat" && powershell'`
>
> For more information on configuring environment variables, see [Use the setvars Script with Linux* or macOS*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html) or [Use the setvars Script with Windows*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).

### On Linux*

1. Change to the sample directory.
2. Configure the build system for the Agilex™ 7 device family, which is the default.

   ```
   mkdir build
   cd build
   cmake ..
   ```

   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ``` 

>    >
>    > Alternatively, you can target an explicit FPGA board variant and BSP by using the following command: 
>    >  ```
>    >  cmake .. -DFPGA_DEVICE=<board-support-package>:<board-variant> -DIS_BSP=1
>    >  ``` 
>    >
>    > You will only be able to run an executable on the FPGA if you specified a BSP.

Reviewer: ```suggestion
```
no full-system support. i won't comment on this anymore, as you can pick up the pattern :)
